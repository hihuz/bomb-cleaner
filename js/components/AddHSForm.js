import React from 'react';
import { connect } from 'react-redux';
import { addHighScore } from '../actions/actionCreators';

class AddHSForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: ''
    };
    this.updateName = this.updateName.bind(this);
    this.handleAddHSClick = this.handleAddHSClick.bind(this);
  }
  updateName(e) {
    this.setState({ name: e.target.value });
  }
  handleAddHSClick() {
    const date = new Date().toJSON().slice(0, 10);
    this.props.dispatch(addHighScore(
      this.props.mode,
      this.state.name,
      this.props.time,
      date,
      this.props.highScores
    ));
    this.props.closeDialog();
  }
  render() {
    return (
      <div className="hs-form">
        You made a high score for <em>{this.props.mode}</em> mode !
        <br /><br />
        Enter your name below :
        <input type="text" onChange={this.updateName} />
        <button className="hs-validate" onClick={this.handleAddHSClick}>OK</button>
      </div>
    );
  }
}

export default connect()(AddHSForm);
