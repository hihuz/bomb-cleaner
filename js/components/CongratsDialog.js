import React from 'react';
import AddHSForm from './AddHSForm';

export function isHighScore(mode, time, highScores) {
  const worseTimes = highScores[mode].filter(hs => time <= hs.time).length;
  return highScores[mode].length < 5 || worseTimes > 0;
}

const CongratsDialog = ({ closeDialog, time, mode, highScores }) => (
  <div>
    <button className="dialog-close" onClick={closeDialog}>
      <i className="icon-close" />
    </button>
    <p className="congrats-para">
      Congratulations !
    </p>
    {mode && time && highScores && isHighScore(mode, time, highScores) ? <AddHSForm
      mode={mode}
      time={time}
      highScores={highScores}
      closeDialog={closeDialog}
    /> : ''}
  </div>
);

export default CongratsDialog;
