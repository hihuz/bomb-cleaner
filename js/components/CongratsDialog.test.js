import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import AddHSForm from './AddHSForm';
import CongratsDialog, { isHighScore } from './CongratsDialog';

const emptyHS = { easy: [], medium: [], hard: [] };
const fullEasyHS = {
  easy: [
    { time: 10 },
    { time: 11 },
    { time: 12 },
    { time: 97 },
    { time: 98 }
  ],
  medium: [],
  hard: []
};

test('Snapshot', () => {
  const component = shallow(<CongratsDialog />);
  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});

test('Should render a AddHSForm component when props give a new valid high score', () => {
  const component = shallow(<CongratsDialog
    mode={'easy'}
    time={70}
    highScores={emptyHS}
  />);
  const actual = component.find(AddHSForm).length;
  expect(actual).toEqual(1);
});

test('Should NOT render a AddHSForm component when props give an invalid high score', () => {
  const component = shallow(<CongratsDialog
    mode={'easy'}
    time={999}
    highScores={fullEasyHS}
  />);
  const actual = component.find(AddHSForm).length;
  expect(actual).toEqual(0);
});

test('Should call the closeDialog prop when the dialog-close button is clicked', () => {
  const clickFn = jest.fn();
  const component = shallow(<CongratsDialog closeDialog={clickFn} />);
  component.find('button').simulate('click');
  expect(clickFn).toHaveBeenCalledTimes(1);
});

test('isHighScore (empty): should return true', () => {
  const actual = isHighScore('easy', 99, emptyHS);
  expect(actual).toEqual(true);
});

test('isHighScore (full / worse score): should return false', () => {
  const actual = isHighScore('easy', 99, fullEasyHS);
  expect(actual).toEqual(false);
});

test('isHighScore (not full / worse score): should return true', () => {
  const hs = Object.assign({}, emptyHS, { easy: [
    { time: 10 },
    { time: 11 },
    { time: 97 },
  ] });
  const actual = isHighScore('easy', 99, hs);
  expect(actual).toEqual(true);
});

test('isHighScore (full / better score): should return true', () => {
  const actual = isHighScore('easy', 96, fullEasyHS);
  expect(actual).toEqual(true);
});

test('isHighScore (not full / better score): should return true', () => {
  const hs = Object.assign({}, emptyHS, { easy: [
    { time: 10 },
    { time: 11 },
    { time: 50 },
    { time: 98 },
  ] });
  const actual = isHighScore('easy', 97, hs);
  expect(actual).toEqual(true);
});
