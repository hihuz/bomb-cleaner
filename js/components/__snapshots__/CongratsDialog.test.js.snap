exports[`test Snapshot 1`] = `
<div>
  <button
    className="dialog-close">
    <i
      className="icon-close" />
  </button>
  <p
    className="congrats-para">
    Congratulations !
  </p>
</div>
`;
